Bluemix: starting point

mbolivar opened the pull request:
This series is based off of work by @mike-scott, with additional patches on top.

It adds basic Bluemix support:

- connect through a gateway
- publish faked temperature data
- optionally register as a managed device

Bluemix works alongside hawkBit as a device manager. That is, data are published to Bluemix regardless of whether hawkBit is used to download firmware.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Application Build Failed.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Application Build Succeeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Nitrogen Deployment Failed.

gatek33p3r commented:
K64F Deployment Failed.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

gatek33p3r commented:
K64F Deployment Succeeded.

mbolivar commented:
Note that this depends on Zephyr patches by @mike-scott currently at https://github.com/mike-scott/zephyr/commits/v1.7-dev.

EmbeddedAndroid commented:
retest

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

mike-scott commented on src/config.h:
```diff
@@ -10,6 +10,12 @@
+ * Bluemix support is built in regardless of this setting , e.g. for
+ * use publishing sensor data.
+ */
 #define CONFIG_DM_BACKEND		BACKEND_HAWKBIT
```
I wonder if we shouldn't add another config like CONFIG_PUBLISH_BLUEMIX or  or something and have the ability to disable Bluemix.  Thinking ahead to AWS and Azure.

mike-scott commented on src/main.c:
```diff
@@ -143,6 +154,31 @@ static void fota_service(void)
+			}
+		} else {
+			/* TODO publish a real temperature */
+			ret = bluemix_pub_temp_c(&bluemix_context, 23);
```
This will need to be decoupled so that it works when CONFIG_DM_BACKEND != BACKEND_BLUEMIX

mike-scott commented on src/tcp.c:
```diff
@@ -44,12 +52,15 @@
 #define PEER_IPADDR		CONFIG_NET_SAMPLES_PEER_IPV4_ADDR
 #endif
+#if (CONFIG_DM_BACKEND == BACKEND_BLUEMIX)
+#define SERVER_PORT		BLUEMIX_PORT
```
How will the BLUEMIX_PORT be used for publish only behavior?  This will need shifting around.

mike-scott commented on src/bluemix_dm.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef __FOTA_BLUEMIX_DM_H__
+#define __FOTA_BLUEMIX_DM_H__
+
+#define BLUEMIX_IPADDR	"fe80::d4e7:0:0:1"
```
We're going to need the BLUEMIX_IPADDR and BLUEMIX_PORT info even when device management isn't handled by Bluemix

mbolivar commented:
v1: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v1

Changes since v1:

- fold in additional patches which allow simultaneous Bluemix and hawkBit connections
- add patch from @mike-scott  to reduce memory footprint by consuming less network-related resources
- add additional fixup patches to beginning of series that were part of the cleanups for Bluemix+hawkBit

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Application Build Succeeded.

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+
+static void connect_cb(struct mqtt_ctx *ctx)
+{
+	OTA_DBG("CONNECT_CB\n");
```
Mind having mqtt as part of the function name or when printing the debug message? It's quite generic now, can be confusing when reading out of this context.

mbolivar replied:
Will do here and elsewhere.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+
+static void disconnect_cb(struct mqtt_ctx *ctx)
+{
+	OTA_DBG("DISCONNECT_CB\n");
```
Same as above.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+static int publish_tx_cb(struct mqtt_ctx *ctx, uint16_t pkt_id,
+			 enum mqtt_packet type)
+{
+	OTA_DBG("PUBLISH_TX_CB\n");
```
Same as above.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+		OTA_ERR("Bluemix topic buffer size %u overflowed by %u B\n",
+			sizeof(bm_ctx->bm_topic),
+			msg->topic_len - sizeof(bm_ctx->bm_topic));
+
```
Extra empty line.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+	OTA_DBG("PUBLISH_RX_CB\n");
+	bm_ctx = mqtt_to_bluemix(ctx);
+
+	if (msg->topic_len > sizeof(bm_ctx->bm_topic)) {
```
Checking for > but you need an extra byte later for \0.

Same when checking for msg_len.

mbolivar replied:
Whoops, nice catch.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+	if (bm_ctx->bm_fatal_err) {
+		/* Propagate fatal error to waiter. */
+		k_sem_give(&bm_ctx->reply_sem);
+		return 0;
```
Ret 0 on fatal errors?

mbolivar replied:
Yup.

We return 0 from the MQTT callback so MQTT keeps going (this is a bluemix fatal error, not an MQTT fatal error). The fatal error is stored in bm_ctx->bm_fatal_err, where it will get picked up by the waiter on bm_ctx->reply_sem.

rsalveti replied:
Great, fine then :-)

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+	int ret;
+
+	while (i++ < APP_CONNECT_TRIES && !ctx->connected) {
+		OTA_DBG(">>> connect: client_id=[%s/%d] user_name=[%s/%d] password=[%s/%d]\n",
```
Should we print the password here?

mike-scott replied:
I was going to ask that this be removed.  I added it when I was confirming the MQTT packaging.

mbolivar replied:
Shh.. don't tell anyone, but the password is just the chip ID number. We need better secret provisioning eventually. But I will take this out.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+
+static int subscribe_to_topic(struct bluemix_ctx *ctx)
+{
+	const char* topics[] = {ctx->bm_topic};
```
s/char* topics/char *topics[]/g
Also give at least one white space after '{' and before '}'. Check for other similar patterns.

mbolivar replied:
Are you saying you want:
```
const char *topics[][]
```

It's already `char *topics[]` above.

I'll fix the space.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+		OTA_ERR("mqtt_tx_subscribe: %d\n", ret);
+		return ret;
+	}
+	ret = wait_for_mqtt(ctx, K_MSEC(1000));
```
Can we have a define for K_MSEC(1000) here?

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+
+static int become_managed_device(struct bluemix_ctx *ctx)
+{
+	int ret;
```
Give one extra line here.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+	if (ret) {
+		return ret;
+	}
+	ret = wait_for_mqtt(ctx, K_MSEC(1200)); /* 400 msec has been observed */
```
Mind adding a define for K_MSEC(1200) as well?

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+#endif	/* CONFIG_BLUEMIX_DM_BACKEND == BACKEND_BLUEMIX */
+
+/* Currently, FOTA app doesn't ever "end" */
+#if 0
```
If not used, just remove for now.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+	OTA_DBG("try_to_connect %d\n", ret);
+	if (ret) {
+		tcp_cleanup(TCP_CTX_BLUEMIX, true);
+		return ret;
```
Either use goto out or remove goto to use return instead. Avoid using 2 different patterns here.

rsalveti commented on src/bluemix.c:
```diff
@@ -0,0 +1,391 @@
+
+int bluemix_pub_temp_c(struct bluemix_ctx *ctx, int temperature)
+{
+	struct mqtt_publish_msg *pub_msg = &ctx->pub_msg;
```
Add extra line here.

rsalveti commented on src/bluemix.h:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2016 Linaro Limited
```
s/2016/2017/g

rsalveti commented on src/bluemix.h:
```diff
@@ -0,0 +1,97 @@
+#elif defined(CONFIG_SOC_SERIES_STM32F4X)
+#define CONFIG_BLUEMIX_DEVICE_TYPE	"carbon"
+#elif defined(CONFIG_SOC_SERIES_KINETIS_K6X)
+#define CONFIG_BLUEMIX_DEVICE_TYPE	"frdm-k64f"
```
Can't we just use the board name used by Zephyr here instead?

rsalveti commented on src/bluemix.h:
```diff
@@ -0,0 +1,97 @@
+	uint8_t bm_auth_token[20]; /* Bluemix authentication token */
+	uint8_t client_id[50];	   /* MQTT client ID */
+
+#define BM_UUID_LEN 36
```
Move this define to the start of the file, avoid mixing declarations.

rsalveti commented on src/bluemix.h:
```diff
@@ -0,0 +1,97 @@
+};
+
+int bluemix_init(struct bluemix_ctx *ctx);
+//int bluemix_uninit(struct bluemix_ctx *ctx);
```
If not used, remove for now.

rsalveti commented on src/config.h:
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright (c) 2016 Linaro Limited
```
s/2016/2017/g

rsalveti commented on src/device.h:
```diff
@@ -81,3 +81,6 @@ extern struct device *flash_dev;
 void set_bluetooth_led(bool state);
+
+/* TODO: Use the proper reboot Zephyr API instead (enabled via Kconfig) */
+extern void sys_arch_reboot(int type);
```
This is now removed from the master branch, so fine to just drop this patch.

rsalveti commented on src/main.c:
```diff
@@ -130,18 +144,47 @@ static void fota_service(void)
+		if (bluemix_failures == MAX_SERVER_FAIL) {
+			printk("Too many bluemix errors, rebooting!\n");
+			/* FIXME: use sys_reboot(). */
+			sys_arch_reboot(0);
```
Just use sys_reboot(0) instead, that was just integrated at master.

rsalveti commented on src/tcp.h:
```diff
@@ -4,15 +4,28 @@
+struct net_context *tcp_get_net_context(enum tcp_context_id id);
+struct k_sem *tcp_get_recv_wait_sem(enum tcp_context_id id);
+
+#endif
```
Add extra info to the endif, like /* __FOTA_TCP_H__ */.

mbolivar replied:
Will do here and in other patches where include guards were added.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Application Build Succeeded.

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

mbolivar commented:
v1: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v1
v2: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v2

Changes since v1:

- fold in additional patches which allow simultaneous Bluemix and hawkBit connections
- add patch from @mike-scott to reduce memory footprint by consuming less network-related resources
- add additional fixup patches to beginning of series that were part of the cleanups for Bluemix+hawkBit

Changes since v2:

- Fix copyright dates (Ricardo)
- Delete unnecessary printk.h include (Ricardo)
- Add MQTT to debug logs for mqtt callbacks (Ricardo)
- Don't overflow message or topic buffers on publish RX (Ricardo)
- Various whitespace and formatting fixes (Ricardo)
- Don't print passwords! Jeez. (Ricardo)
- Use defines for timeouts (Ricardo)
- Remove unused Bluemix uninit routine (Ricardo)
- Use consistent error handling in bluemix_init() (Ricardo)
- Add comments for MQTT_MSG_MAX_SIZE choice (Ricardo)
- Use CONFIG_BOARD for CONFIG_BLUEMIX_DEVICE_TYPE (Ricardo)
- Use sys_reboot() instead of sys_arch_reboot() (Ricardo)

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Application Build Succeeded.

mbolivar commented:
v1: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v1
v2: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v2
v3: https://github.com/mbolivar/zephyr-fota-hawkbit/tree/bluemix-wip-v3

Changes since v1:

- fold in additional patches which allow simultaneous Bluemix and hawkBit connections
- add patch from @mike-scott to reduce memory footprint by consuming less network-related resources
- add additional fixup patches to beginning of series that were part of the cleanups for Bluemix+hawkBit

Changes since v2:

- Fix copyright dates (Ricardo)
- Delete unnecessary printk.h include (Ricardo)
- Add MQTT to debug logs for mqtt callbacks (Ricardo)
- Don't overflow message or topic buffers on publish RX (Ricardo)
- Various whitespace and formatting fixes (Ricardo)
- Don't print passwords! Jeez. (Ricardo)
- Use defines for timeouts (Ricardo)
- Remove unused Bluemix uninit routine (Ricardo)
- Use consistent error handling in bluemix_init() (Ricardo)
- Add comments for MQTT_MSG_MAX_SIZE choice (Ricardo)
- Use CONFIG_BOARD for CONFIG_BLUEMIX_DEVICE_TYPE (Ricardo)
- Use sys_reboot() instead of sys_arch_reboot() (Ricardo)

Changes since v3:

- Seriously, don't overflow message or topic buffers (Ricardo)

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

rsalveti commented:
LGTM, thanks for the updates. Not yet tested, Mike can test and do the final +1.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Nitrogen Deployment Failed.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Application Build Succeeded.

mbolivar commented:
Please hold off reviewing v5 -- seeing odd errors in my logs.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Application Build Succeeded.

gatek33p3r commented:
K64F Deployment Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

gatek33p3r commented:
Checkpatch Suceeded.

gatek33p3r commented:
Bootloader Build Succeeded.

gatek33p3r commented:
Application Build Succeeded.

gatek33p3r commented:
Nitrogen Deployment Succeeded.

gatek33p3r commented:
K64F Deployment Failed.
